docs(editor): KB for adding icons (#2300)
telerik/blazor-docs: knowledge-base/editor-add-icons.md

---
title: Add Icons in Editor

> description: How to insert icon tags in Editor?

Reviewer: Insert a longer description (between 100-150 characters).

type: how-to
page_title: Add Icons in Editor
slug: editor-kb-add-icons
position: 
tags: editor, blazor, icon, schema, prosemirror
ticketid: 
res_type: kb
---

## Environment
<table>
	<tbody>
		<tr>
			<td>Product</td>
			<td>Editor for Blazor</td>
		</tr>
	</tbody>
</table>

## Description

I want to add custom icons in the Editor content. When adding an icon such as `<i class="fa-light fa-envelope">`, the icon's HTML disappears upon saving.

> How can I add an `<i>` element inside the Editor?

Reviewer: This sounds like an alternative question describing the same request. Consider adding "This Knowledge Base article also answers the following question:".

## Solution

> The built-in ProseMirror schema that the Editor uses does not support `<i>` tag. To allow adding icons in the Editor content:

Reviewer: ```suggestion
The built-in ProseMirror schema that the Editor uses does not support an `<i>` tag. To allow adding icons in the Editor content:
```

1. [Modify the ProseMirror schema]({%slug editor-modify-default-schema%}) to include a node for the `<i>` element.

> 2. Ensure the CSS classes for the icons are applied. The approach varies depending on the [edit mode of the Editor]({%slug editor-edit-modes-overview%})

Reviewer: ```suggestion
2. Ensure the CSS classes for the icons are applied. The approach varies depending on the [edit mode of the Editor]({%slug editor-edit-modes-overview%}):
```

    * [Iframe Edit Mode](#add-icons-in-iframe-edit-mode)
    * [Div Edit Mode](#add-icons-in-div-edit-mode)

> ### Add icons in Iframe Edit Mode

Reviewer: ```suggestion
### Add Icons in Iframe Edit Mode
```

When the [EditMode of the Editor is set to `EditorEditMode.Iframe`]({%slug editor-edit-modes-iframe%}), the content area is an editable `<iframe>` element that does not inherit the CSS rules from the current page.

> This means that you need to inject the icons stylesheet into the `<iframe>`, so the icons are properly rendered. [This will be supported in a future UI for Blazor version](https://feedback.telerik.com/blazor/1543925-add-the-ability-to-inject-css-files-into-the-iframe). For the time-being you can inject the stylesheet with JSInterop in the `OnAfterRenderAsync`.

Author: The IFrame is the default edit mode and I believe a lot of people use it. That said, I'd prefer to give the customers an option they can use now as it may be a while before we implement the feature. We even used this approach in the docs - https://docs.telerik.com/blazor-ui/components/editor/prosemirror-plugins#adding-a-custom-plugin.
I realize this part will be outdated when we release the feature but this is why I've linked the feature request there. I may add a not indicating the version the kb was written in.

> 

Reviewer: Consider listing the main steps covered by the code example. You can do this either by using a list before the example or as comments in the code snippet. This allows the reader to grasp that major milestones in the proposed implementation quickly.
This comment applies to the next snippet as well.

            },
        ],
        toDOM: (node) => {
            const attrs = {
                class: node.attrs.class
            };
            return ["i", attrs];
        },
    };

    window.schemaProvider = (args) => {
        const schema = args.getSchema();
        const Schema = args.ProseMirror.Schema

        let nodes = schema.spec.nodes.addToEnd("i", icon);

        const newSchema = new Schema({ nodes });
        return newSchema;
    }

    function injectEditorStylesheet() {
        var doc = document.querySelector("iframe").contentWindow.document;
        var head = doc.querySelector("head");

        var cssLink1 = document.createElement("link");
        cssLink1.href = "https://cdnjs.cloudflare.com/ajax/libs/open-iconic/1.1.1/font/css/open-iconic-bootstrap.min.css";
        cssLink1.rel = "stylesheet";
        cssLink1.type = "text/css";

        var cssLink2 = document.createElement("link");
        cssLink2.href = "https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/all.min.css";
        cssLink2.rel = "stylesheet";
        cssLink2.type = "text/css";

        head.appendChild(cssLink1);
        head.appendChild(cssLink2);
    };
</script>
````

> ### Add icons in Div Edit Mode

Reviewer: ```suggestion
### Add Icons in Div Edit Mode
```

@code {
    private string EditorValue { get; set; } = @"Here is an example icon in the Editor content <i class='fa fa-info-circle'></i>";

    private List<IEditorTool> EditorTools { get; set; } = new List<IEditorTool>() { new ViewHtml() };
}

@* Move JavaScript code to a separate JS file in production *@
<script suppress-error="BL9992">
    var icon = {
        attrs: {
            class: { default: null },
            type: { default: null },
        },
        group: "inline",
        content: "text*",
        inline: true,
        parseDOM: [
            {
                tag: "i",
                getAttrs: (dom) => ({
                    class: dom.getAttribute("class"),
                }),
            },
        ],
        toDOM: (node) => {
            const attrs = {
                class: node.attrs.class
            };
            return ["i", attrs];
        },
    };
    debugger

    window.schemaProvider = (args) => {
        const schema = args.getSchema();
        const Schema = args.ProseMirror.Schema

        let nodes = schema.spec.nodes.addToEnd("i", icon);

        const newSchema = new Schema({ nodes });
        return newSchema;
    }

</script>

> ```

Reviewer: Consider adding a See Also section with a few relevant links.

Author: See https://github.com/telerik/blazor-docs/pull/2297#discussion_r1704107336

Reviewer 2: Some scenarios may require users to add icons and not just see them. So, you can link here to the Custom Tools article, which is not mentioned in the KB itself.

---

> title: Add Icons in Editor

Reviewer 2: The focus is font icons (and not images), so we can clarify this upfront.

```suggestion
title: Add Font Icons in Editor
```

description: How to allow inserting <i> tags in the Editor for Blazor to render custom icons in the Editor content?
type: how-to
page_title: Add Icons in Editor
slug: editor-kb-add-icons
position: 

> tags: editor, blazor, icon, schema, prosemirror

Reviewer 2: ```suggestion
tags: blazor, editor, prosemirror, schema, icon
```

ticketid: 
res_type: kb
---

## Environment
<table>
	<tbody>
		<tr>
			<td>Product</td>
			<td>Editor for Blazor</td>
		</tr>
	</tbody>
</table>

## Description

This KB article answers the following questions:

* How to add custom icons in the Editor content? When adding an icon such as `<i class="fa-light fa-envelope">`, the icon's HTML disappears upon saving.
* How can I add an `<i>` element inside the Editor?

## Solution

> The built-in ProseMirror schema that the Editor uses does not support an `<i>` tag. To allow adding icons in the Editor content:

Reviewer 2: This statement implies that the Editor schema depends on ProseMirror and is outside our control, which is not the case. Consider something like:

```suggestion
The default schema of the Telerik Editor does not include an `<i>` tag, so the Editor strips such elements automatically. To allow adding font icons through `<i>` tags in the Editor content:
```

1. [Modify the ProseMirror schema]({%slug editor-modify-default-schema%}) to include a node for the `<i>` element.

> 2. Ensure the CSS classes for the icons are applied. The approach varies depending on the [edit mode of the Editor]({%slug editor-edit-modes-overview%}):

Reviewer 2: Our customers know that. The tricky part that is in our support scope is how to inject the stylesheet. So:

```suggestion
2. Ensure the required font icon stylesheets can affect the Editor content. The approach varies depending on the [edit mode of the Editor]({%slug editor-edit-modes-overview%}):
```

    * [Iframe Edit Mode](#add-icons-in-iframe-edit-mode)
    * [Div Edit Mode](#add-icons-in-div-edit-mode)

### Add Icons in Iframe Edit Mode

> When the [EditMode of the Editor is set to `EditorEditMode.Iframe`]({%slug editor-edit-modes-iframe%}), the content area is an editable `<iframe>` element that does not inherit the CSS rules from the current page.

Reviewer 2: ```suggestion
When the [Editor `EditMode` is set to `EditorEditMode.Iframe`]({%slug editor-edit-modes-iframe%}), the content area is inside an `<iframe>` element that does not apply the CSS rules from the current page.
```

This means that you need to inject the icons stylesheet into the `<iframe>`, so the icons are properly rendered. At the time of writing (UI for Blazor **6.1.0**), [the Editor does not support injecting your CSS files into the Iframe](https://feedback.telerik.com/blazor/1543925-add-the-ability-to-inject-css-files-into-the-iframe) but you can inject them with JSInterop in the `OnAfterRenderAsync`.

>caption Add icons in an Editor with Iframe Edit Mode

````CSHTML
@using Telerik.Blazor.Components.Editor
@inject IJSRuntime js

<TelerikEditor @bind-Value="@EditorValue"
               Tools="@EditorTools"
               Schema="schemaProvider"
               Width="600px"
               Height="300px">
</TelerikEditor>

@code {

>     private string EditorValue { get; set; } = @"Here is an example icon in the Editor content <i class='fa fa-info-circle'></i>";

Reviewer 2: Not sure if the Editor will add `<p>` tags automatically, but it is more realistic to add them to the initial value ourselves.

```suggestion
    private string EditorValue { get; set; } = @"<p>Here is an example icon in the Editor content <i class='fa fa-info-circle'></i></p>";
```

    private List<IEditorTool> EditorTools { get; set; } = new List<IEditorTool>() { new ViewHtml() };

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            await js.InvokeVoidAsync("injectEditorStylesheet");
        }
    }
}

@* Move JavaScript code to a separate JS file in production *@
<script suppress-error="BL9992">

    //define the icon node

>     var icon = {

Reviewer 2: Consider a more descriptive variable name.

```suggestion
    var iconNode = {
```

        content: "text*",
        inline: true,
        parseDOM: [
            {
                tag: "i",
                getAttrs: (dom) => ({
                    class: dom.getAttribute("class"),
                }),
            },
        ],
        toDOM: (node) => {
            const attrs = {
                class: node.attrs.class
            };
            return ["i", attrs];
        },
    };

    //add the icon node to the Editor ProseMirror schema
    window.schemaProvider = (args) => {
        const schema = args.getSchema();
        const Schema = args.ProseMirror.Schema

        let nodes = schema.spec.nodes.addToEnd("i", icon);

        const newSchema = new Schema({ nodes });
        return newSchema;
    }

    //inject the stylesheet for the icons, so they are properly visualized
    function injectEditorStylesheet() {
        var doc = document.querySelector("iframe").contentWindow.document;
        var head = doc.querySelector("head");

        var cssLink1 = document.createElement("link");
        cssLink1.href = "https://cdnjs.cloudflare.com/ajax/libs/open-iconic/1.1.1/font/css/open-iconic-bootstrap.min.css";
        cssLink1.rel = "stylesheet";

>         cssLink1.type = "text/css";

Reviewer 2: I think the `type` attribute is optional for stylesheets.

        var cssLink2 = document.createElement("link");
        cssLink2.href = "https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/all.min.css";
        cssLink2.rel = "stylesheet";
        cssLink2.type = "text/css";

        head.appendChild(cssLink1);
        head.appendChild(cssLink2);
    };

</script>
````

### Add Icons in Div Edit Mode

When the [EditMode of the Editor is set to `EditorEditMode.Div`]({%slug editor-edit-modes-div%}), the content area is an editable <div> element that inherits the CSS rules from the current page.

> This allows you to include the icons stylesheet in the `<head>` of the web page along with the other stylesheets. 

Reviewer 2: Noun adjectives should be singular.

```suggestion
This allows you to include the icon stylesheets in the `<head>` of the web page along with the other stylesheets. 
```

> >caption Add icons in an Editor with Div Edit Mode

Reviewer 2: Either "EditMode" or "edit mode"

        fontAwesomeCssLink.rel = "stylesheet";

        head.appendChild(bootstrapCssLink);
        head.appendChild(fontAwesomeCssLink);
    };

</script>
````

### Add Icons in Div Edit Mode

When the [Editor `EditMode` is set to `EditorEditMode.Div`]({%slug editor-edit-modes-div%}), the content area is a `<div contenteditable="true">` element that inherits the CSS rules from the current page.

This allows you to include the icon stylesheets in the `<head>` of the web page along with the other stylesheets. 

>caption Add icons in an Editor with Div edit mode

````CSHTML
@using Telerik.Blazor.Components.Editor

@* Just one example of including custom font icon libraries.
Make sure to use the correct way and resources for your actual project *@
<link href="https://cdnjs.cloudflare.com/ajax/libs/open-iconic/1.1.1/font/css/open-iconic-bootstrap.min.css" rel="stylesheet" />
<link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/all.min.css" rel="stylesheet" />

<TelerikEditor @bind-Value="@EditorValue"
               Tools="@EditorTools"
               Schema="schemaProvider"
               EditMode="@EditorEditMode.Div"

>                Width="600px"

Author: ```suggestion
```

---
title: Add Font Icons in Editor

> description: How to allow inserting <i> tags in the Editor for Blazor to render custom icons in the Editor content?

Reviewer 2: ```suggestion
description: How to insert <i> tags in the Editor for Blazor to render custom icons in the Editor content?
```

type: how-to

> page_title: Add Icons in Editor

Reviewer 2: ```suggestion
page_title: How to Add Font Icons in the Blazor Editor
```

			<td>Product</td>
			<td>Editor for Blazor</td>
		</tr>
	</tbody>
</table>

## Description

This KB article answers the following questions:

* How to add custom icons in the Editor content? When adding an icon such as `<i class="fa-light fa-envelope">`, the icon's HTML disappears upon saving.
* How can I add an `<i>` element inside the Editor?

## Solution

The default schema of the Telerik Editor does not include an `<i>` tag, so the Editor strips such elements automatically. To allow adding font icons through `<i>` tags in the Editor content:

1. [Modify the ProseMirror schema]({%slug editor-modify-default-schema%}) to include a node for the `<i>` element.
2. Ensure the required font icon stylesheets can affect the Editor content. The approach varies depending on the [edit mode of the Editor]({%slug editor-edit-modes-overview%}):
    * [Iframe Edit Mode](#add-icons-in-iframe-edit-mode)
    * [Div Edit Mode](#add-icons-in-div-edit-mode)

### Add Icons in Iframe Edit Mode

When the [Editor `EditMode` is set to `EditorEditMode.Iframe`]({%slug editor-edit-modes-iframe%}), the content area is inside an `<iframe>` element that does not apply the CSS rules from the current page.

> This means that you need to inject the icons stylesheet into the `<iframe>`, so the icons are properly rendered. At the time of writing (UI for Blazor **6.1.0**), [the Editor does not support injecting your CSS files into the Iframe](https://feedback.telerik.com/blazor/1543925-add-the-ability-to-inject-css-files-into-the-iframe) but you can inject them with JSInterop in the `OnAfterRenderAsync`.

Reviewer 2: ```suggestion
This means that you need to inject the icons stylesheet into the `<iframe>`, so the icons are properly rendered. At the time of writing (UI for Blazor **6.1.0**), [the Editor does not support injecting your CSS files into the iframe](https://feedback.telerik.com/blazor/1543925-add-the-ability-to-inject-css-files-into-the-iframe) but you can inject them with JSInterop in `OnAfterRenderAsync`.
```

>caption Add icons in an Editor with Iframe edit mode

````CSHTML
@using Telerik.Blazor.Components.Editor
@inject IJSRuntime js

<TelerikEditor @bind-Value="@EditorValue"
               Tools="@EditorTools"
               Schema="schemaProvider"

>                Width="600px"

Reviewer 2: The Editor `Width` may be larger than the width of the REPL box, e.g. on phones. We can skip it unless necessary for some reason.

>                Width="600px"

Author: ```suggestion
```

    private List<IEditorTool> EditorTools { get; set; } = new List<IEditorTool>() { new ViewHtml() };
}

@* Move JavaScript code to a separate JS file in production *@
<script suppress-error="BL9992">

    //define the icon node
    var iconNode = {
        attrs: {
            class: { default: null },
            type: { default: null },
        },
        group: "inline",
        content: "text*",
        inline: true,
        parseDOM: [
            {
                tag: "i",
                getAttrs: (dom) => ({
                    class: dom.getAttribute("class"),
                }),
            },
        ],
        toDOM: (node) => {
            const attrs = {
                class: node.attrs.class
            };
            return ["i", attrs];
        },
    };
    debugger

    //add the icon node to the Editor ProseMirror schema
    window.schemaProvider = (args) => {
        const schema = args.getSchema();
        const Schema = args.ProseMirror.Schema

        let nodes = schema.spec.nodes.addToEnd("i", iconNode);

        const newSchema = new Schema({ nodes });
        return newSchema;
    }

</script>
````

## See Also

> * [Editor Custom Tools]({%slug editor-custom-tools%})

Reviewer 2: ```suggestion
* [Custom Editor Tools]({%slug editor-custom-tools%})
```
